apps/etl-func/src/adapters/functions/messages-ingestion.ts

```diff
@@ -1,42 +1,55 @@
+import { EventErrorRepository } from "@/domain/event.js";
 import { IngestMessageUseCase } from "@/domain/use-cases/ingest-message.js";
-import { CosmosDBHandler } from "@azure/functions";
-import { pino } from "pino";
+import { CosmosDBHandler, InvocationContext } from "@azure/functions";
 
 import {
   MessageMetadata,
   messageMetadataSchema,
 } from "../../domain/message.js";
 
-const logger = pino({
-  level: process.env.NODE_ENV === "production" ? "error" : "debug",
-});
-
-const processMessageMetadata = (
-  input: unknown,
-): MessageMetadata | undefined => {
-  const result = messageMetadataSchema.safeParse(input);
-  //we are considering, and returning, only processed messages. A processed message has isPending === false
-  if (result.success && !result.data.isPending) return result.data;
-  else return undefined;
-};
-
 const messagesIngestionHandler =
-  (ingestUseCase: IngestMessageUseCase): CosmosDBHandler =>
-  async (documents: unknown[]) => {
-    //Avoiding all documents different from MessageMetadata schema and with
-    //isPending equals to true
-
-    const parsedMessagesMetadata = documents.map(processMessageMetadata);
-    const messagesMetadata: MessageMetadata[] = parsedMessagesMetadata.filter(
-      (item): item is MessageMetadata => item !== undefined,
-    );
+  (
+    ingestUseCase: IngestMessageUseCase,
+    eventErrorRepository: EventErrorRepository<unknown>,
+  ): CosmosDBHandler =>
+  async (documents: unknown[], context: InvocationContext) => {
+    const parsedMessagesMetadata: MessageMetadata[] = [];
+    const invalidDocuments: unknown[] = [];
 
+    documents.forEach((document) => {
+      const result = messageMetadataSchema.safeParse(document);
+      if (!result.success) {
+        invalidDocuments.push(document);
+        return;
+      }
+      if (!result.data.isPending) {
+        parsedMessagesMetadata.push(result.data);
+      }
+    });
+    let success = false;
```

Comment:
I would avoid this pattern in order to respect the principle of immutable data in functional programming. I am aware that this approach requires higher computational costs, but I would prefer something like
```suggestion
    const parsedMessagesMetadataOrZodError = documents.map((input) => {
    const parsedInput = messageMetadataSchema.safeParse(input);
    if (parsedInput.success) return parsedInput.data;
    else return parsedInput.error;
  })
  
  // filter the array in order to get only valid MessageMetadata with
  // isPending = false
  const nonPendingMessageMetadata = parsedMessagesMetadataOrZodError.filter(
    (input): input is MessageMetadata =>
      !(input instanceof ZodError) && !input.isPending,
  );

  // get all malformed documents so we can log them
  const malformedDocuments = parsedMessagesMetadataOrZodError.filter(
    (input): input is ZodError => input instanceof ZodError,
  );    
```
This way you can pass `nonPendingMessageMetadata` to the use case.
What's your take? @lucamancosu

Reply:
Yes i like to use the same paradigm we have used in the ingestion process. @acuiuli was worried about the computational costs, and ingestion time cost, so we have did a try pushing the code like that.

I know about the functional requirement and there is no problem for me. @acuiuli What do you think ?

Reply:
We did it like this as Luca said to have it more optimized since the computational cost is already high. If we prefer to do it in a more functional way I can modify it.

Reply:
modified [here](https://github.com/pagopa/io-messages/pull/86/commits/649dafa6aaff455943f9fa4e5a330859996ff197)